Up sve 3

hzhuang1 opened the pull request:
Implement arm64 dispatch with assembly SVE routines.

easyaspi314 committed:
Change dispatch breakpoint to XXH3_accumulate()

This is in preparation for an SVE implementation

easyaspi314 committed:
Fix errors

hzhuang1 committed:
debug: fix empty target

Signed-off-by: Haojian Zhuang <haojian.zhuang@linaro.org>

hzhuang1 committed:
xxhash: support SVE by intrinsic code

Implement intrinsic code for AARCH64 SVE. It could get the performance
improvement from 40% to 210% for different large inputs.

It seems that accessing memory cost a lot of resources by SVE instructions.
The acc array is only 512-bit long that could easily loaded into Z registers.
Performance could be improved if not access the acc array from memory
frequently.

SCALAR implementation (default)
benchmarking large inputs : from 512 bytes (log9) to 128 MB (log27)
xxh3   ,  2587,  3015,  3152,  3296,  3383,  3425,  3449,  3425,  3431,  3457,  3467,  3467,  3466,  3254,  2400,  2151,  2134,  2116,  2119
XXH32  ,  1498,  1524,  1537,  1544,  1543,  1546,  1548,  1536,  1491,  1490,  1493,  1493,  1490,  1441,  1242,  1186,  1181,  1184,  1188
XXH64  ,  2630,  2874,  3017,  3092,  3131,  3142,  3154,  3130,  3013,  3020,  3021,  3017,  3017,  2870,  2034,  1834,  1853,  1854,  1842
XXH128 ,  2353,  2711,  3057,  3244,  3352,  3409,  3443,  3435,  3431,  3441,  3396,  3441,  3431,  3224,  2299,  2125,  2145,  2164,  2192

intrinsic SVE512 implementation with prefetch instruction
benchmarking large inputs : from 512 bytes (log9) to 128 MB (log27)
xxh3   ,  3684,  5967,  7804,  8945,  9866, 10406, 10691, 10656, 10854, 10925, 10922, 10949, 10960, 10029,  6347,  6034,  5958,  5991,  6059
XXH32  ,  1326,  1440,  1495,  1523,  1534,  1543,  1544,  1535,  1505,  1507,  1508,  1506,  1507,  1460,  1257,  1199,  1189,  1192,  1198
XXH64  ,  2510,  2803,  2976,  3067,  3120,  3138,  3155,  3130,  3051,  3058,  3056,  3063,  3055,  2932,  2163,  1990,  1962,  1971,  1981
XXH128 ,  3421,  5798,  7500,  8881,  9790, 10348, 10669, 10592, 10855, 10968, 11021, 11029, 11020,  9988,  6156,  5821,  5937,  5994,  6079

Signed-off-by: Haojian Zhuang <haojian.zhuang@linaro.org>
Signed-off-by: Devin Hussey <easyaspi314@users.noreply.github.com>

hzhuang1 commented:
It's a new version of #683 
In this version, both intrinsic code and assembly code are included.

hzhuang1 commented:
The first two commits come from Devin. And the third commit is used to fix the CI errors.
Let's review the last 3 commits.

hzhuang1 committed:
add prefetch instrution in assembly SVE routine

Use prefetch instruction to tune the performance.

benchmarking large inputs : from 512 bytes (log9) to 128 MB (log27)
xxh3   ,  4040,  6610,  9690, 12158, 14054, 15201, 15891, 15506, 15384, 13975, 14045, 14059, 14070, 11838,  4572,  3953,  3997,  3999,  3990
XXH32  ,  1321,  1439,  1494,  1521,  1534,  1541,  1546,  1534,  1502,  1505,  1493,  1504,  1505,  1444,  1249,  1195,  1195,  1191,  1195
XXH64  ,  2502,  2799,  2975,  3070,  3115,  3136,  3152,  3127,  3040,  3038,  3051,  3052,  3036,  2857,  2142,  1975,  1980,  1980,  1979
XXH128 ,  3842,  6341,  9479, 11898, 13938, 15005, 15819, 15245, 15278, 15505, 15630, 15668, 15712, 12437,  4912,  4148,  4020,  4028,  4010

Signed-off-by: Haojian Zhuang <haojian.zhuang@linaro.org>

hzhuang1 commented:
@Cyan4973 @easyaspi314 How about this patch set? Could you have any comments to share? Thanks

hzhuang1 commented:
> @Cyan4973 @easyaspi314 How about this patch set? Could you have any comments to share? Thanks

@Cyan4973 @easyaspi314 What should I do for next? Thanks

Cyan4973 commented:
Hi @hzhuang1 

I don't think you need to do anything else.
It's mostly a matter of time, we need to find enough availability to make a proper review of your code.
You did as much and as well as you could up to now.
Now the ball is on our side.

Cyan4973 commented on Makefile:
```diff
@@ -97,6 +97,11 @@ all: lib xxhsum xxhsum_inlinedXXH
 xxhsum: CPPFLAGS += -DXXHSUM_DISPATCH=1
 xxhsum: xxh_x86dispatch.o
+DISPATCH_HEADERS = xxh_x86dispatch.h
+else ifeq ($(ARM64_DISPATCH),1)
```
`ARM64_DISPATCH` provides a way to manually set dispatch for ARM64.
This works if `DISPATCH` is not set, and needs to be documented, otherwise it's unlikely to be discovered.

Another alternative could be to automate `ARM64` target detection.
This would require calling the compiler with its intended set of compilation flags, and make it tell which platform it's targeting, then trigger `x86` or `arm64` dispatch depending on the detected platform, or even opt out if detection fails or feels fishy. 
Ideally, it should be possible to override automated detection with manual settings, just in case, and also to satisfy requirements from package managers.

I understand this is additional nice-to-have, and makes more sense to invest into it once the main patch is accepted.

hzhuang1 replied:
I rechecked the Makefile. Maybe it could be implemented by `uname -p` command. The command result is in below.
| Command | Mac on x86 | Mac on arm | Linux on x86 | Linux on arm |
|---------------| ---------------: | ---------------: | -----------------: | ----------------: |
| uname -p  | i386             | arm              | x86_64           | aarch64        | 

And I need to make sure that Makefile is only used in MacOS/Linux/UNIX in this project. CMAKE is used for Windows OS. Is it right? If so, I could create a macro to distinguish architecture.

Cyan4973 replied:
Sounds like a plan

Cyan4973 commented on xxh_arm64dispatch.h:
```diff
@@ -0,0 +1,86 @@
+extern "C" {
+#endif
+
+XXH64_hash_t  XXH3_64bits_dispatch(const void* input, size_t len);
```
This seems to be exactly the same interface contract as `xxh_x86dispatch.h` ?
In which case, maybe it's better to combine them ?

hzhuang1 replied:
There's a bit difference. In `xxh_x86dispatch.h`, the interfaces are defined with XXH_PUBLIC_API. I removed these for `xxh_arm64dispatch.h` since it would result in some build warnings.
```
clang -O3 -march=armv8-a+sve -fPIC -Wall -Wextra -Wstrict-aliasing=1 -std=c99     -DARM64_DISPATCH=1 -I../..     -c -o main.o main.c
In file included from main.c:36:
In file included from ./hashes.h:68:
./../../xxh_arm64dispatch.h:45:30: warning: function 'XXH3_64bits_dispatch' has internal linkage but is not defined [-Wundefined-internal]
XXH_PUBLIC_API XXH64_hash_t  XXH3_64bits_dispatch(const void* input, size_t len);
                             ^
./hashes.h:91:21: note: used here
    return (size_t) XXH3_64bits(src, srcSize);
                    ^
./../../xxh_arm64dispatch.h:65:22: note: expanded from macro 'XXH3_64bits'
# define XXH3_64bits XXH3_64bits_dispatch
                     ^
./../../xxh_arm64dispatch.h:50:30: warning: function 'XXH3_128bits_dispatch' has internal linkage but is not defined [-Wundefined-internal]
XXH_PUBLIC_API XXH128_hash_t XXH3_128bits_dispatch(const void* input, size_t len);
                             ^
./hashes.h:98:21: note: used here
    return (size_t) XXH3_128bits(src, srcSize).low64;
                    ^
./../../xxh_arm64dispatch.h:76:23: note: expanded from macro 'XXH3_128bits'
# define XXH3_128bits XXH3_128bits_dispatch
                      ^
2 warnings generated.
```
I'm not sure whether I could remove these for x86. So I just created a copy with a few differences.

Cyan4973 replied:
It's not clear to me why there are these differences compared to the `x86` dispatcher.
Would be worth an investigation.

Cyan4973 commented on xxh_arm64sve.S:
```diff
@@ -0,0 +1,475 @@
+
+	.arch armv8-a+sve
+
+#include "asmdefs.h"
```
This is an assembly file, not a C file with some assembly.
It means the compilation pipeline will be different for this file.

This may look like a minor detail, but I've experienced critical issues doing the same thing in other projects.

Probably the most critical one is the impact this can have on the executable stack flag, which then trigger security related issues.
It needs to be taken care of at compiler level.
This can be an annoyance to setup. You may refer to https://github.com/facebook/zstd/pull/2964/files#diff-a19342636d0d39fbbfc6ebf4e3839ed2ae4e043e7c9e37ca96226c0b7ad3a499R101 for an example.

hzhuang1 replied:
Does it mean that I could avoid executable stack flag if I don't include `asmdefs.h`?

Cyan4973 replied:
> Does it mean that I could avoid executable stack flag if I don't include `asmdefs.h`?

No, the issue is about `xxh_arm64sve.S` itself

Cyan4973 commented on xxhash.h:
```diff
@@ -4861,20 +5077,20 @@ XXH3_hashLong_internal_loop(xxh_u64* XXH_RESTRICT acc,
         {   const xxh_u8* const p = input + len - XXH_STRIPE_LEN;
 #define XXH_SECRET_LASTACC_START 7  /* not aligned on 8, last secret is different from acc & scrambler */
-            f_acc512(acc, p, secret + secretSize - XXH_STRIPE_LEN - XXH_SECRET_LASTACC_START);
+            XXH3_accumulate_512(acc, p, secret + secretSize - XXH_STRIPE_LEN - XXH_SECRET_LASTACC_START);
```
Question :
how can you guarantee that `XXH3_accumulate_512()` will trigger the right variant
in the context of dynamic dispatch ?

hzhuang1 replied:
Thanks for catching this issue. I should define `XXH3_accumulate_512()` to `XXH3_accumulate_512_scalar()` for ARM64_DISPATCH case. Since it's only directly invoked for the last strip.

Cyan4973 commented:
Could you document, in the PR, the performance implications of this patch ?

Cyan4973 commented:
I was wondering if this PR could be made smaller by separating 2 components : 
1) introduce a new `SVE` micro-op support (triggered through static selection only)
2) then introduce a dynamic dispatcher for `arm64`

Unless I'm missing something, it seems this could help to review each part specifically.

Cyan4973 commented:
Summary : it looks like great work,
but it also looks like a lot of code to add, 
so we have to be cautious about its potential side effects,
both for existing user base, and for future maintenance.

Anything that can help make this PR smaller will be welcome.
Hence the suggestion to split it into 2 parts.

It may also trigger deeper questions about the repository structure.
The core library `xxhash` was built around the requirement of being a "single file library" (not counting here the CLI, nor the dispatcher).
This PR seems to change that requirement, by introducing an additional assembly file for `SVE`.
And of course, more files for the dispatcher.

So it's a direct contributor to increased repository complexity.
This could nonetheless be worth it, but this complexity must be brought under control, in order to keep this repository usable for "casual" users that just want to grab `"xxhash.h"` and maybe `"xxhash.c"` and call it a day.
I think it's still possible for non `SVE` users, but it's less and less clear that it is, so even discoverability shall be taken seriously, both at documentation and repository organization levels.

I don't claim to have all answers right now, but these are the topics to consider to make this contribution part of mainline (as opposed to be maintained into its own fork, where of course such requirements can be relaxed).

hzhuang1 commented:
> Could you document, in the PR, the performance implications of this patch ?

Sure. I'll document it. The problem is that only performance of small data is recorded in `README.md`. My optimization is mainly focus on large block data. Which documentation should I use? or just fill them in the comments of the PR?

hzhuang1 commented:
> I was wondering if this PR could be made smaller by separating 2 components :
> 
>     1. introduce a new `SVE` micro-op support (triggered through static selection only)
> 
>     2. then introduce a dynamic dispatcher for `arm64`
> 
OK. I'll split it into two patches.

Cyan4973 commented:
> > Could you document, in the PR, the performance implications of this patch ?
> 
> Sure. I'll document it. The problem is that only performance of small data is recorded in `README.md`. My optimization is mainly focus on large block data. Which documentation should I use? or just fill them in the comments of the PR?

Yes, in the PR, that's fine

hzhuang1 committed:
dispatch different implementation on arm64 by macro

Dispatch SVE, NEON and SCALAR implementations on arm64 by selecting
different macros.

Since SVE implementation can't be supported by compiler well, use assembly
code instead.

In intrinsic SVE implementation:
  * Avoid to access the ACC array frequently in accumulation routine.

In assembly SVE implementation (dispatcher):
  * Avoid to access the ACC array frequently in accumulation routine.
  * Use assemly code in scramble routine.
  * Since there's both accumulation and scramble routine in internal loop,
    convert the internal loop to assembly version. At this time, avoid to
    access the ACC array frequently in the internal loop.

Signed-off-by: Haojian Zhuang <haojian.zhuang@linaro.org>

hzhuang1 committed:
use assembly SVE routine in tests

Make bench tests to support assembly SVE routine.

  $export CPP_FLAGS="-DXXH_VECTOR=XXH_SVE"
  $export CFLAGS="-O3 -march=armv8-a+sve -fPIC -DXXH_VECTOR=XXH_SVE"
  $make DISPATCH=1

benchmarking large inputs : from 512 bytes (log9) to 128 MB (log27)
xxh3   ,  3978,  6534,  9666, 12155, 14084, 15239, 15853, 13357, 11540, 11203, 11242, 11255, 11274,  9845,  4989,  4568,  4490,  4525,  4564
XXH32  ,  1325,  1440,  1495,  1521,  1534,  1543,  1546,  1532,  1491,  1492,  1492,  1494,  1492,  1447,  1238,  1189,  1178,  1184,  1188
XXH64  ,  2508,  2804,  2965,  3054,  3119,  3130,  3148,  3121,  3002,  3008,  3010,  3013,  3011,  2869,  2105,  1971,  1940,  1950,  1966
XXH128 ,  3835,  6318,  9463, 11976, 13984, 15190, 15794, 13514, 11491, 11611, 11699, 11699, 11725, 10593,  5138,  4658,  4461,  4517,  4578

Signed-off-by: Haojian Zhuang <haojian.zhuang@linaro.org>

hzhuang1 committed:
dynamic dispatch on aarch64

User could avoid to select XXH_VECTOR any more. If environment
variable DISPATCH is enabled, dispatcher could select right method
automatically.

  $export CFLAGS="-O3 -march=armv8-a+simd -fPIC"
  $make DISPATCH=1

Signed-off-by: Haojian Zhuang <haojian.zhuang@linaro.org>

hzhuang1 committed:
set GNU stack note for noexecstack

If an assembler source contains no GNU-stack note, the system by default
assumes that an executable stack may be required. GCC generates code to
be executed on the stack when it implements a trampoline for nested
functions. The default behavior brings out security issue.

Signed-off-by: Haojian Zhuang <haojian.zhuang@linaro.org>

hzhuang1 committed:
share same dispatch head file

Share same dispatch head file between x86_64 and aarch64 platform, since
the interfaces are same.

Signed-off-by: Haojian Zhuang <haojian.zhuang@linaro.org>

hzhuang1 commented:
Close it since new pull request (https://github.com/Cyan4973/xxHash/pull/713) is created.
